### PR TITLE
[pipeline] Bert pipeline for shardformer and its tests

### PR DESCRIPTION
**Pull request by CjhHa1**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`
#4190 

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.
This PR add bert pipeline into shardformer and tests the pipeline forward.

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

**Comment by CjhHa1** [comment]
截屏2023-07-10 10 38 39

截屏2023-07-10 10 39 02

shard former tests and pipeline tests passed

File: colossalai/shardformer/policies/bert.py [hunk 1]
```diff
@@ -176,6 +176,13 @@ class BertModelPolicy(BertPolicy):
     def __init__(self) -> None:
         super().__init__()
 
+    def module_policy(self):
+        module_policy = super().module_policy()
+        from transformers.models.bert.modeling_bert import BertModel
+        module_policy[BertModel] = ModulePolicyDescription(
+            method_replacement={'forward': partial(bert_model_forward, stage_manager=self.pipeline_stage_manager)})
```
Comment (FrankLeeeee) [inline colossalai/shardformer/policies/bert.py]: Should check if pipeline parallelism is enabled?

File: tests/kit/model_zoo/torchrec/__init__.py [hunk 2]
```diff
@@ -1 +1 @@
-from .torchrec import *
+#from .torchrec import *
```
Comment (FrankLeeeee) [inline tests/kit/model_zoo/torchrec/__init__.py:1]: Dont' comment out in your commit, only do this for your local testing.

File: colossalai/shardformer/policies/bert.py (hunk 1 above)
Comment (CjhHa1) [reply colossalai/shardformer/policies/bert.py]: I am sorry that I did not push the updated code in time, please see the new commit!

File: tests/kit/model_zoo/torchrec/__init__.py (hunk 2 above)
Comment (CjhHa1) [reply tests/kit/model_zoo/torchrec/__init__.py:1]: done

File: tests/test_shardformer/test_model/test_shard_bert_pipeline.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+from tests.test_shardformer.test_model._utils import build_model, build_pipeline_model, run_forward
+
+
+def check_forward_backward(org_model, sharded_model, data_gen_fn, output_transform_fn, loss_fn):
+    # check forward
+    pass
+
+
+@parameterize('enable_fused_normalization', [False])
+@parameterize('enable_tensor_parallelism', [False])
```
Comment (FrankLeeeee) [inline tests/test_shardformer/test_model/test_shard_bert_pipeline.py:26]: Should add pipeline parallelism to parameterize as well?

File: colossalai/shardformer/policies/bert.py (hunk 1 above)
Comment (FrankLeeeee) [reply colossalai/shardformer/policies/bert.py]: Good!

File: tests/kit/model_zoo/torchrec/__init__.py (hunk 2 above)
Comment (FrankLeeeee) [reply tests/kit/model_zoo/torchrec/__init__.py:1]: It is still commented out.

File: tests/test_shardformer/test_model/test_shard_bert_pipeline.py (hunk 3 above)
Comment (CjhHa1) [reply tests/test_shardformer/test_model/test_shard_bert_pipeline.py:26]: In this test the pipeline is always on.

File: tests/kit/model_zoo/torchrec/__init__.py (hunk 2 above)
Comment (CjhHa1) [reply tests/kit/model_zoo/torchrec/__init__.py]: It's alright now, check it

File: colossalai/shardformer/shard/sharder.py [hunk 4]
```diff
@@ -134,7 +135,9 @@ def _replace_param(
     def _replace_method(self, module: nn.Module, method_replacement: Dict[str, Callable]):
         for method_name, new_method in method_replacement.items():
             # bind the new method to the module
-            setattr(module, method_name, new_method.__get__(module, module.__class__))
+            bound_method = MethodType(new_method, module)
+            original_method = getattr(module, method_name)
```
Comment (ver217) [inline colossalai/shardformer/shard/sharder.py:139]: Where do you use this?

File: tests/test_shardformer/test_model/_utils.py [hunk 5]
```diff
@@ -16,6 +17,24 @@ def build_model(model_fn, enable_fused_normalization=True, enable_tensor_paralle
     return org_model, sharded_model.cuda()
 
 
+def build_pipeline_model(model_fn,
+                         stage_manager=None,
+                         enable_fused_normalization=False,
+                         enable_tensor_parallelism=False):
+    # create new model
+    org_model = model_fn().cuda()
```
Comment (ver217) [inline tests/test_shardformer/test_model/_utils.py:25]: Please add lazy init.

File: tests/test_shardformer/test_model/test_shard_bert_pipeline.py (hunk 3 above)
Comment (FrankLeeeee) [reply tests/test_shardformer/test_model/test_shard_bert_pipeline.py:26]: Ok

File: colossalai/shardformer/shard/sharder.py (hunk 4 above)
Comment (CjhHa1) [reply colossalai/shardformer/shard/sharder.py:139]: I am sorry, this should be discarded

File: tests/test_shardformer/test_model/_utils.py (hunk 5 above)
Comment (CjhHa1) [reply tests/test_shardformer/test_model/_utils.py]: done

File: colossalai/shardformer/shard/sharder.py (hunk 4 above)
Comment (CjhHa1) [reply colossalai/shardformer/shard/sharder.py]: done

**Review by ver217 — APPROVED** [review]

**Comment by CjhHa1** [comment]
截屏2023-07-10 13 39 59
截屏2023-07-10 13 57 42

finish tests

**Pull request merged**